src/daft-local-execution/src/progress_bar.rs

```diff
@@ -0,0 +1,84 @@
+use std::{borrow::Cow, sync::atomic::AtomicU64};
+
+use indicatif::{HumanCount, ProgressBar, ProgressStyle};
+
+pub enum ProgressBarColor {
+    Blue,
+    Magenta,
+    Cyan,
+}
+
+impl ProgressBarColor {
+    fn to_str(&self) -> &'static str {
+        match self {
+            Self::Blue => "blue",
+            Self::Magenta => "magenta",
+            Self::Cyan => "cyan",
+        }
+    }
+}
+
+pub struct OperatorProgressBar {
+    inner_progress_bar: ProgressBar,
+    emitted: AtomicU64,
+    show_received: bool,
+}
+
+impl OperatorProgressBar {
+    pub fn new(
+        prefix: impl Into<Cow<'static, str>>,
+        color: ProgressBarColor,
+        show_received: bool,
+    ) -> Self {
+        let template_str = if show_received {
+            format!(
+                "🗡️ 🐟 {{spinner:.green}} {{prefix:.{color}/bold}} | [{{elapsed_precise}}] {{human_pos}} rows received, {{msg}} rows emitted",
+                color = color.to_str(),
+            )
+        } else {
+            format!(
+                "🗡️ 🐟 {{spinner:.green}} {{prefix:.{color}/bold}} | [{{elapsed_precise}}] {{msg}} rows emitted",
+                color = color.to_str(),
+            )
+        };
+
+        let pb = ProgressBar::new_spinner()
+            .with_style(
+                ProgressStyle::default_spinner()
+                    .template(template_str.as_str())
+                    .unwrap(),
+            )
+            .with_prefix(prefix)
+            .with_message("0".to_string());
+        Self {
+            inner_progress_bar: pb,
+            emitted: AtomicU64::new(0),
+            show_received,
+        }
+    }
+
+    pub fn increment_received(&self, amount: u64) {
+        if self.show_received {
+            self.inner_progress_bar.inc(amount);
+        }
+    }
+
+    pub fn increment_emitted(&self, amount: u64) {
+        let prev = self
+            .emitted
+            .fetch_add(amount, std::sync::atomic::Ordering::Relaxed);
+        let count = prev + amount;
+        self.inner_progress_bar
```

Comment:
I ran this using a fast in memory query I think this is happening too fast.

For `set_progress` and `inc`, the progress bar is using a rate limiter. You don't want to redraw the progress bar on every single update but only if enough time has passed.

Take a look at https://github.com/console-rs/indicatif/blob/6417492caa546a91ca6f69881d33141ddaf192c2/src/progress_bar.rs#L233

Reply:
Nice find, changed it so that it updates only every 100ms. Also changed it so that it updates both rows received and rows emitted through a single `set_message` (instead of both `inc` and `set_message`).